### PR TITLE
feat: aot badges

### PR DESCRIPTION
**Pull request by ayrock-dev**
## Description

Implement badges for Advent of TypeScript.

:bulb: Note: badges are *not* db persisted. Badges are calculated on page load from challenge data in server action.

I've implemented 3 badges:
|Badge Name|Criteria|
|---|---|
|Bronze|In progress on > 0 AOT challenges (aka user has started at least one, but has not finished even one).|
|Silver|Completed 5+ AOT challenges.|
|Gold|Completed 15+ AOT challenges.|
|Gold|Completed all 25 AOT challenges.|

## Related Issue

Part of #1120

## Screenshots/Video (if applicable)

image

If user has not earned any badges, badges section is hidden:

image
<br />

### CODE REVIEW

**Comment by bautistaaa** [comment]
we should try those badges @zaCKoZAck0 posted in discord. super sexy

**Comment by zaCKoZAck0** [comment]
I think I can cook smth.

**Comment by ayrock-dev** [comment]
aot badge mocks
https://www.figma.com/file/haGIxp448IBoYePtLZuRqp

image

**Comment by ayrock-dev** [comment]
note this does NOT put completed number on top of badges. we can iterate there if we want.

File: apps/web/src/app/[locale]/(profile)/[username]/_components/dashboard/_actions.ts [hunk 1]
```diff
@@ -152,3 +152,81 @@ export async function getSolvedChallenges() {
...
+          id: true,
+        },
+      },
+    },
+  });
+
+  const numberOfInProgressHolidayChallenges = holidayTrack.trackChallenges.filter(
+    (trackChallenge) => {
+      return (
+        trackChallenge.challenge.submission.length &&
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/(profile)/[username]/_components/dashboard/_actions.ts:200]: do you have to check length here or can you just `?.every`

File: apps/web/src/app/[locale]/(profile)/[username]/_components/dashboard/_actions.ts [hunk 2]
```diff
@@ -152,3 +152,81 @@ export async function getSolvedChallenges() {
...
+  ).length;
+
+  if (numberOfInProgressHolidayChallenges > 0) {
+    badges.push({ slug: 'aot-2023-bronze', name: 'Advent of TypeScript 2023 Bronze' });
+  }
+
+  const numberOfCompletedHolidayChallenges = holidayTrack.trackChallenges.filter(
+    (trackChallenge) => {
+      return (
+        trackChallenge.challenge.submission.length &&
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/(profile)/[username]/_components/dashboard/_actions.ts:213]: do you have to check length here or can you just ?.some

**Review by bautistaaa — APPROVED** [review]

File: apps/web/src/app/[locale]/(profile)/[username]/_components/dashboard/_actions.ts (hunk 1 above)
Comment (ayrock-dev) [reply apps/web/src/app/[locale]/(profile)/[username]/_components/dashboard/_actions.ts]: sounds good, implemented. im not much of a `?.` fan in boolean logic (feels "gotcha-y" compared to a length check) but happy to conform.

**Review by zaCKoZAck0 — APPROVED** [review]

**Pull request merged**